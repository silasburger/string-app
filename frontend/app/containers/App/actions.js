/*
 *
 * StringPage actions
 *
 */

import {
  FETCH_POSTS,
  POSTS_RECEIVED,
  FETCH_POSTS_FAILED,
  POST_CREATED,
  CREATE_POST_FAILED,
  CREATE_POST,
} from './constants';

export function fetchPosts() {
  return {
    type: FETCH_POSTS,
  };
}
export function fetchPostsFailed(error) {
  return {
    type: FETCH_POSTS_FAILED,
    payload: error,
  };
}
export function postsReceived(posts) {
  return {
    type: POSTS_RECEIVED,
    payload: { posts },
  };
}
export function createPost() {
  return {
    type: CREATE_POST,
  };
}
export function createPostFailed(error) {
  return {
    type: CREATE_POST_FAILED,
    payload: error,
  };
}
export function postCreated(post) {
  return {
    type: POST_CREATED,
    payload: post,
  };
}
