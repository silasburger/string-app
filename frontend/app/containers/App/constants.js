/*
 * AppConstants
 * Each action has a corresponding type, which the reducer knows and picks up on.
 * To avoid weird typos between the reducer and the actions, we save them as
 * constants here. We prefix them with 'yourproject/YourComponent' so we avoid
 * reducers accidentally picking up actions they shouldn't.
 *
 * Follow this format:
 * export const YOUR_ACTION_CONSTANT = 'yourproject/YourContainer/YOUR_ACTION_CONSTANT';
 */

export const FETCH_POSTS = 'app/App/FETCH_POSTS';
export const POSTS_RECEIVED = 'app/App/POSTS_RECEIVED';
export const FETCH_POSTS_FAILED = 'app/App/FETCH_POSTS_FAILED';
export const CREATE_POST = 'app/App/CREATE_POST';
export const POST_CREATED = 'app/App/POST_CREATED';
export const CREATE_POST_FAILED = 'app/App/CREATE_POST_FAILED';
