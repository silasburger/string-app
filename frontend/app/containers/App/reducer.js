/*
 * AppReducer
 *
 * The reducer takes care of our data. Using actions, we can change our
 * application state.
 * To add a new action, add it to the switch statement in the reducer function
 *
 * Example:
 * case YOUR_ACTION_CONSTANT:
 *   return state.set('yourStateVariable', true);
 */

import { fromJS } from 'immutable';

import {
  FETCH_POSTS,
  POSTS_RECEIVED,
  FETCH_POSTS_FAILED,
  POST_CREATED,
  CREATE_POST_FAILED,
  CREATE_POST,
} from './constants';

// The initial state of the App
const initialState = fromJS({
  loading: false,
  error: false,
  posts: false,
});

function appReducer(state = initialState, action) {
  switch (action.type) {
    case FETCH_POSTS:
      console.log('fetching posts');
      return state.set('loading', true).set('error', false);
    case POSTS_RECEIVED:
      console.log('posts received');
      return state
        .set('loading', false)
        .set('error', false)
        .set('posts', action.payload.posts);
    case FETCH_POSTS_FAILED:
      return state.set('loading', false).set('error', action.payload.error);
    case CREATE_POST:
      return state.set('loading', true).set('error', false);
    case CREATE_POST_FAILED:
      return state.set('loading', false).set('error', action.payload.error);
    case POST_CREATED:
      return state.set('loading', false).set('error', false);
    default:
      return state;
  }
}

export default appReducer;
