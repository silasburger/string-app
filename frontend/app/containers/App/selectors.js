/**
 * The global state selectors
 */

import { createSelector } from 'reselect';

const selectRouter = state => state.get('router');

const selectGlobal = state => state.get('global');

const makeSelectLocation = () =>
  createSelector(selectRouter, routerState =>
    routerState.get('location').toJS(),
  );

const makeSelectLoading = () =>
  createSelector(selectGlobal, globalState =>
    globalState.get('loading').toJS(),
  );

const makeSelectError = () =>
  createSelector(selectGlobal, globalState => globalState.get('error').toJS());

export { makeSelectLocation, makeSelectError, makeSelectLoading };
