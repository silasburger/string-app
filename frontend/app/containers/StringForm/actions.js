/*
 *
 * StringForm actions
 *
 */

import {
  CHANGE_STRING,
  CREATE_POST,
  CREATE_POST_FAILED,
  POST_CREATED,
} from './constants';

export function changeString(value) {
  return {
    type: CHANGE_STRING,
    value,
  };
}
export function createPost(data) {
  return {
    type: CREATE_POST,
    data,
  };
}
export function createPostFailed(error) {
  return {
    type: CREATE_POST_FAILED,
    error,
  };
}
export function postCreated(post) {
  return {
    type: POST_CREATED,
    post,
  };
}
