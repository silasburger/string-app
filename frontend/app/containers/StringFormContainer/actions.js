/*
 *
 * StringFormContainer actions
 *
 */

import {
  CHANGE_STRING,
  CREATE_POST,
  CREATE_POST_FAILED,
  POST_CREATED,
} from './constants';

export function changeString(value) {
  return {
    type: CHANGE_STRING,
    value,
  };
}
export function createPost() {
  return {
    type: CREATE_POST,
  };
}
export function createPostFailed(error) {
  return {
    type: CREATE_POST_FAILED,
    error,
  };
}
export function postCreated(post) {
  return {
    type: POST_CREATED,
    post,
  };
}
