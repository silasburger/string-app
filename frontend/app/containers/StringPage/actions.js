/*
 *
 * StringPage actions
 *
 */

import { FETCH_POSTS, POSTS_RECEIVED, FETCH_POSTS_FAILED } from './constants';

export function fetchPosts() {
  console.log('in fetch posts action');
  return {
    type: FETCH_POSTS,
  };
}
export function fetchPostsFailed(error) {
  return {
    type: FETCH_POSTS_FAILED,
    error,
  };
}
export function postsReceived(posts) {
  return {
    type: POSTS_RECEIVED,
    posts,
  };
}
