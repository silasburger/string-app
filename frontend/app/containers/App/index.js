/**
 *
 * App.js
 *
 * This component is the skeleton around the actual pages, and should only
 * contain code that should be seen on all pages. (e.g. navigation bar)
 *
 */

import React from 'react';
import { Switch, Route } from 'react-router-dom';
import { Helmet } from 'react-helmet';

import StringPage from 'containers/StringPage/Loadable';
import FormPage from 'components/FormPage';
import NotFoundPage from 'containers/NotFoundPage/Loadable';
import NavBar from '../../components/NavBar';

import GlobalStyle from '../../global-styles';

export default function App() {
  return (
    <div>
      <Helmet>
        <title>DMI Tech Screen</title>
        <meta name="DMI Tech Screen" content="An initial tech screen" />
        <noscript>Javascrirpt must be enabled for this site to work</noscript>
      </Helmet>
      <NavBar />
      <Switch>
        <Route exact path="/" component={FormPage} />
        <Route exact path="/strings" component={StringPage} />
        <Route component={NotFoundPage} />
      </Switch>
      <GlobalStyle />
    </div>
  );
}
