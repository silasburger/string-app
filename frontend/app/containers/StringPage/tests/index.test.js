import React from 'react';
import { mount, shallow } from 'enzyme';

import { fetchPosts } from 'containers/App/actions';
import { changeDateTimeFilter } from '../actions';
import { StringPage, mapDispatchToProps } from '../index';

describe('<StringPage />', () => {
  it('string page snapshot', () => {
    const wrapper = shallow(
      <StringPage
        loading={false}
        error={false}
        posts={[]}
        filter="ALL_POSTS"
        changeDateTimeFilter={() => {}}
        fetchPosts={() => {}}
      />,
    ).dive();
    expect(wrapper).toMatchSnapshot();
  });

  it('should fetch posts on mount', () => {
    const fetchSpy = jest.fn();
    mount(
      <StringPage
        loading={false}
        error={false}
        posts={[]}
        filter="ALL_POSTS"
        changeDateTimeFilter={() => {}}
        fetchPosts={fetchSpy}
      />,
    );
    expect(fetchSpy).toHaveBeenCalled();
  });
  describe('mapDispatchToProps', () => {
    describe('fetchPosts', () => {
      it('should be injected', () => {
        const dispatch = jest.fn();
        const result = mapDispatchToProps(dispatch);
        expect(result.fetchPosts).toBeDefined();
      });

      it('should dispatch fetchPosts when called', () => {
        const dispatch = jest.fn();
        const result = mapDispatchToProps(dispatch);
        result.fetchPosts();
        expect(dispatch).toHaveBeenCalledWith(fetchPosts());
      });
    });

    describe('changeDateTimeFilter', () => {
      it('should be injected', () => {
        const dispatch = jest.fn();
        const result = mapDispatchToProps(dispatch);
        expect(result.changeDateTimeFilter).toBeDefined();
      });

      it('should dispatch change when called', () => {
        const dispatch = jest.fn();
        const result = mapDispatchToProps(dispatch);
        result.changeDateTimeFilter({ target: { value: 'ALL_POSTS' } });
        expect(dispatch).toHaveBeenCalledWith(
          changeDateTimeFilter('ALL_POSTS'),
        );
      });
    });
  });
});
