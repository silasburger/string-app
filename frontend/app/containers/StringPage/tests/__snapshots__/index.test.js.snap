// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<StringPage /> string page snapshot 1`] = `
<div
  id="string-page"
>
  <HelmetWrapper
    defer={true}
    encodeSpecialCharacters={true}
  >
    <title>
      String Page
    </title>
    <meta
      content="This page shows all the strings!"
      name="String Page"
    />
  </HelmetWrapper>
  <LoadableComponent />
  <PostList
    changeDateTimeFilter={[Function]}
    filter="ALL_POSTS"
    loading={false}
    posts={Array []}
  />
</div>
`;
